recon/src/recon.rs

```diff
@@ -21,114 +25,141 @@ use tracing::{instrument, trace};
 /// Recon is generic over its Key, Hash and Store implementations.
 /// This type provides the core protocol implementation.
 #[derive(Debug)]
-pub struct Recon<K, H, S>
+pub struct Recon<K, H, S, I>
 where
     K: Key,
     H: AssociativeHash,
     S: Store<Key = K, Hash = H>,
+    I: InterestProvider<Key = K>,
 {
+    interests: I,
     store: S,
 }
 
-impl<K, H, S> Recon<K, H, S>
+impl<K, H, S, I> Recon<K, H, S, I>
 where
     K: Key,
     H: AssociativeHash,
     S: Store<Key = K, Hash = H>,
+    I: InterestProvider<Key = K>,
 {
     /// Construct a new Recon instance.
-    pub fn new(store: S) -> Self {
-        Self { store }
+    pub fn new(store: S, interests: I) -> Self {
+        Self { store, interests }
     }
+
     /// Construct a message to send as the first message.
-    pub fn initial_message(&self) -> Message<K, H> {
-        let mut response = Message::<K, H>::default();
-        if let Some((first, last)) = self.store.first_and_last(&K::min_value(), &K::max_value()) {
-            let len = self.store.len();
-            match len {
-                0 => {
-                    // this should be unreachable, but we explicitly add this case to be clear.
-                }
-                1 => {
-                    // -> (only_key)
-                    response.keys.push(first.to_owned());
-                }
-                2 => {
-                    // -> (first 0 last)
-                    response.keys.push(first.to_owned());
-                    response.hashes.push(H::identity());
-                    response.keys.push(last.to_owned());
-                }
-                _ => {
-                    // -> (first h(middle) last)
-                    response.keys.push(first.to_owned());
-                    response.hashes.push(self.store.hash_range(first, last));
-                    response.keys.push(last.to_owned());
+    pub fn initial_messages(&self) -> Result<Vec<Message<K, H>>> {
+        let interests = self.interests()?;
+        let mut messages = Vec::with_capacity(interests.len());
+        for range in interests {
+            let mut response = Message::new(&range.start, &range.end);
+            if let Some((first, last)) = self.store.first_and_last(&range.start, &range.end) {
+                // Cannot use store.len, we need the count between first and last
+                let len = self.store.len();
+                match len {
+                    0 => {
+                        // this should be unreachable, but we explicitly add this case to be clear.
```

Comment:
Although this should be unreachable, does it make sense to throw an error here in case dependent code breaks this assumption?

Reply:
In this case, the correct behavior is to do nothing anyway so that is why it doesn't through an error.